gstcefsrc.cc

```diff
@@ -524,67 +656,109 @@ run_cef (GstCefSrc *src)
 
     /* unblock start () */
     g_mutex_lock (&init_lock);
-    cef_inited = TRUE;
-    g_cond_signal(&init_cond);
+    cef_status = CEF_STATUS_FAILURE;
+    g_cond_broadcast (&init_cond);
     g_mutex_unlock (&init_lock);
 
     goto done;
   }
 
   g_mutex_lock (&init_lock);
-  cef_inited = TRUE;
-  init_result = TRUE;
-  g_cond_signal(&init_cond);
+  cef_status = CEF_STATUS_INITIALIZED;
+  g_cond_broadcast (&init_cond);
   g_mutex_unlock (&init_lock);
 
+#ifndef __APPLE__
   CefRunMessageLoop();
-
-  CefShutdown();
-
-  g_mutex_lock (&init_lock);
-  cef_inited = FALSE;
-  g_cond_signal(&init_cond);
-  g_mutex_unlock (&init_lock);
+  gst_cef_shutdown(nullptr);
+#endif
 
 done:
   return NULL;
 }
 
-void quit_message_loop (int arg)
+static GstStateChangeReturn
+gst_cef_src_change_state(GstElement *element, GstStateChange transition)
 {
-  CefQuitMessageLoop();
-}
+  auto result = GST_CALL_PARENT_WITH_DEFAULT (GST_ELEMENT_CLASS , change_state, (element, transition), GST_STATE_CHANGE_FAILURE);
 
-class ShutdownEnforcer {
- public:
-  ~ShutdownEnforcer() {
-    if (!cef_inited)
-      return;
-
-    CefPostTask(TID_UI, base::BindOnce(&quit_message_loop, 0));
-
-    g_mutex_lock(&init_lock);
-    while (cef_inited)
+  switch(transition)
+  {
+  case GST_STATE_CHANGE_NULL_TO_READY:
+  {
+    g_mutex_lock (&init_lock);
+    // Wait till a previous CEF is dismantled or completes initialization
+    while (cef_status & CEF_STATUS_MASK_TRANSITIONING)
       g_cond_wait (&init_cond, &init_lock);
+    if (cef_status == CEF_STATUS_FAILURE) {
+      // BAIL OUT, CEF is not loaded.
+      result = GST_STATE_CHANGE_FAILURE;
+    } else if (cef_status == CEF_STATUS_NOT_LOADED) {
+      cef_status = CEF_STATUS_INITIALIZING;
+      /* Initialize Chromium Embedded Framework */
+#ifdef __APPLE__
+      /* in the main thread as per Cocoa */
+      if (gst_cef_is_main_thread()) {
+        g_mutex_unlock (&init_lock);
+        run_cef ((GstCefSrc*) element);
+        g_mutex_lock (&init_lock);
+      } else {
+        dispatch_async_f(dispatch_get_main_queue(), (GstCefSrc*)element, (dispatch_function_t)&run_cef);
+        while (cef_status == CEF_STATUS_INITIALIZING)
+          g_cond_wait (&init_cond, &init_lock);
+      }
+#else
+        /* in a separate UI thread */
+      thread = g_thread_new("cef-ui-thread", (GThreadFunc) run_cef, (GstCefSrc*)element);
+      while (cef_status == CEF_STATUS_INITIALIZING)
+        g_cond_wait (&init_cond, &init_lock);
+#endif
+      if (cef_status != CEF_STATUS_INITIALIZED) {
+        // BAIL OUT, CEF is not loaded.
+        result = GST_STATE_CHANGE_FAILURE;
+#ifndef __APPLE__
+        g_thread_join(thread);
+        thread = nullptr;
+#endif
+      }
+    }
+    g_mutex_unlock(&init_lock);
+    break;
+  }
+  case GST_STATE_CHANGE_READY_TO_NULL:
+  {
+    g_mutex_lock (&init_lock);
+    if (cef_status == CEF_STATUS_SHUTTING_DOWN) {
```

Comment:
Multiple elements could get into this `if` and then on Apple call `gst_cef_shutdown()`. Should that be avoided? Then you'd have to change `cef_status` before the `g_cond_wait()` to something else

Reply:
Yeah, shouldn't the implementation instead read:

```
      while (cef_status & CEF_STATUS_MASK_TRANSITIONING)
        g_cond_wait (&init_cond, &init_lock);
      /* At that point the element holds the lock and knows no status change is happening */
      if (browsers == 0 && cef_status != CEF_STATUS_NOT_LOADED) {
         /* This element is the one in charge of shutting down */
         cef_status = CEF_STATUS_SHUTTING_DOWN;
         /* Here quit message loop if needed, dispatch cef_shutdown otherwise */
         ...
         /* Now wait for shut down to be over */
         while (cef_status != CEF_STATUS_NOT_LOADED)
              g_cond_wait (&init_cond, &init_lock);
         /* Clean up */
         ...
      }
```

This also means that `BrowserClient::OnBeforeClose` should now only update the browsers count :)

Reply:
Just applied this, with one nit: the check must be against `CEF_STATUS_INITIALIZED`, otherwise CEF asserts because there's not a global context active.

Reply:
You forgot to push this or is github confused?

Reply:
Let me see, I think it got pushed in the wrong repo.